linux/LibSensors.c

```diff
@@ -149,6 +152,62 @@ static int tempDriverPriority(const sensors_chip_name* chip) {
    return -1;
 }
 
+void LibSensors_countCCDs(CPUData* cpus, unsigned int existingCPUs) {
+   /* For AMD k10temp/zenpower, temperatures are provided for CCDs only,
+      which is an aggregate of multiple cores.
+      There's no obvious mapping between hwmon sensors and sockets and CCDs.
+      Assume both are iterated in order.
+      Hypothesis: Each CCD has same size N = #Cores/#CCD
+      and is assigned N coreID in sequence.
+      Also assume all CPUs have same number of CCDs. */
+
+   int ccds = 0;
+
+   for (size_t i = 0; i < existingCPUs + 1; i++) {
+      cpus[i].ccdID = -1;
+   }
+
+   int n = 0;
+   for (const sensors_chip_name* chip = sym_sensors_get_detected_chips(NULL, &n); chip; chip = sym_sensors_get_detected_chips(NULL, &n)) {
+      int m = 0;
+      for (const sensors_feature* feature = sym_sensors_get_features(chip, &m); feature; feature = sym_sensors_get_features(chip, &m)) {
+         if (feature->type != SENSORS_FEATURE_TEMP)
+            continue;
+
+         if (!feature->name || !String_startsWith(feature->name, "temp"))
+            continue;
+
+         char *label = sym_sensors_get_label(chip, feature);
+         if (label) {
+            if (String_startsWith(label, "Tccd")) {
+               ccds++;
+            }
+            free(label);
+         }
+      }
+   }
+
+   if (ccds > 0) {
+      int ccdsPerCore = existingCPUs / ccds;
+
+      int ccd = 0;
+      int nc = ccdsPerCore;
+      for (int p = 0; p < (int)existingCPUs; p++) {
+         for (int c = 0; c < (int)existingCPUs; c++) {
+            for (size_t i = 1; i <= existingCPUs; i++) {
+               if (cpus[i].physicalID == p && cpus[i].coreID == c) {
+                  cpus[i].ccdID = ccd;
+                  if (--nc == 0) {
+                     nc = ccdsPerCore;
+                     ccd++;
+                  }
+               }
+            }
+         }
+      }
+   }
```

Comment:
Try to avoid deep nesting by early returns or negating conditions to use `continue` in strategic places.

Reply:
```suggestion
   if (ccds < 1)
      return;

   int ccdsPerCore = existingCPUs / ccds;

   int ccd = 0;
   int nc = ccdsPerCore;

   for (int p = 0; p < (int)existingCPUs; p++) {
      for (int c = 0; c < (int)existingCPUs; c++) {
         for (size_t i = 1; i <= existingCPUs; i++) {
            if (cpus[i].physicalID != p || cpus[i].coreID != c)
               continue;

            cpus[i].ccdID = ccd;
            if (--nc == 0) {
               nc = ccdsPerCore;
               ccd++;
            }
         }
      }
   }
```

Any reason for this last loop to be O(n³)?

Reply:
This was the most straightforward way to write this, but perhaps I can find a better solution...

Reply:
I think a O(n) option could be:

1. Alloc memory for each CCD to store plus a 2D array of pointers to map the CCD ID. Initialize each entry to -1 (unassigned); index by physical id/core id.
2. Iterate over all CPU cores (loop `i`) and check if `ccd[pid][cid]` has been set: If so, assign the current CPU to the given CCD, otherwise allocate a new one and use that one instead.
3. Free the temporary memory.

Reply:
One could also do one iteration over all CPU to find max core_id and phyiscal_id and then keep the O(n^3) loop with vastly smaller n.

Reply:
Technically that's O(n*m), but yes. Also this check could be included with the reading of the physical and core id values in the first place, making the cost effectively zero. Having these limits also reduces the memory cost of the suggested algorithm from my previous comment.